src/aspire/abinitio/commonline_cn.py

```diff
@@ -0,0 +1,376 @@
+                    f"Only Cn symmetry supported. {symmetry} was supplied."
+                )
+            self.order = int(symmetry[1:])
+
+    def _estimate_relative_viewing_directions(self):
+        n_img = self.n_img
+        n_theta = self.n_theta
+        pf = self.pf
+        max_shift_1d = self.max_shift
+        shift_step = self.shift_step
+
+        # Generate candidate rotation matrices and the common-line and
+        # self-common-line indices induced by those rotations.
+        Ris_tilde, R_theta_ijs = self.generate_cand_rots(
+            self.n_points_sphere,
+            self.equator_threshold,
+            self.order,
+            self.degree_res,
+            self.seed,
+        )
+        cijs_inds = self._compute_cls_inds(Ris_tilde, R_theta_ijs)
+        scls_inds = self._compute_scls_inds(Ris_tilde)
+        n_cands = len(Ris_tilde)
+        n_theta_ijs = len(R_theta_ijs)
+
+        # Generate shift phases.
+        r_max = pf.shape[-1]
+        shifts, shift_phases, _ = self._generate_shift_phase_and_filter(
+            r_max, max_shift_1d, shift_step
+        )
+        n_shifts = len(shifts)
+        all_shift_phases = shift_phases.T
+
+        # Reconstruct full polar Fourier for use in correlation.
+        pf /= norm(pf, axis=2)[..., np.newaxis]  # Normalize each ray.
+        pf_full = np.concatenate((pf, np.conj(pf)), axis=1)
+
+        # Pre-compute shifted pf's.
+        pf_shifted = (pf * all_shift_phases[:, None, None]).swapaxes(0, 1)
+        pf_shifted = pf_shifted.reshape((n_img, n_shifts * (n_theta // 2), r_max))
+
+        # Step 1: pre-calculate the likelihood with respect to the self-common-lines.
+        scores_self_corrs = np.zeros((n_img, n_cands), dtype=self.dtype)
+        logger.info("Computing likelihood wrt self common-lines.")
+        for i in trange(n_img):
+            # Compute correlation of pf[i] with itself over all shifts.
+            corrs = pf_shifted[i] @ np.conj(pf_full[i]).T
+            corrs = np.reshape(corrs, (n_shifts, n_theta // 2, n_theta))
+            corrs_cands = np.max(
+                np.real(corrs[:, scls_inds[:, :, 0], scls_inds[:, :, 1]]), axis=0
+            )
+
+            scores_self_corrs[i] = np.mean(corrs_cands, axis=1)
+
+        # Step 2: Compute the likelihood for each pair of candidate matrices with respect
+        # to the common-lines they induce.
+        logger.info("Computing pairwise likelihood.")
+        n_vijs = n_img * (n_img - 1) // 2
+        vijs = np.zeros((n_vijs, 3, 3), dtype=self.dtype)
+        viis = np.zeros((n_img, 3, 3), dtype=self.dtype)
+        rots_symm = cyclic_rotations(self.order, self.dtype).matrices
+        c = 0
+
+        # List of MeanOuterProductEstimator instances.
+        # Used to keep a running mean of J-synchronized estimates for vii.
+        mean_est = []
+        for _ in range(n_img):
+            mean_est.append(MeanOuterProductEstimator())
+
+        with tqdm(total=n_vijs) as pbar:
```

Comment:
Would it make sense to use `all_pairs` here and just loop over those?

Reply:
Yes, this is much nicer with `all_pairs`. Only one loop deep now :)